File: public/javascripts/src/util/ImmutableData.js

/**
 * 2) Implement an immutable class in JavaScript. The class fully encapsulates an array 
 * of integers, an arbitrary data object, * and a `Date`. It should allow clients to 
 * specify all three when the class is constructed. 
 * Once instantiated, the value of * each property should not change.
 */
export default class ImmutableData {

	/**
	 * Create a new set of immutable data.
	 * 
	 * @param {[number]} numbers A plain array of numbers. 
	 * @param {object} obj 
	 * @param {Date} date 
	 */
	constructor(numbers, obj, date) {

		this._numbers = numbers.map(x => x); // copy using the identity function
		this._obj = obj;
		this._date = date;

		if (Object.freeze) {
			ImmutableData.deepFreeze(this._numbers);
			ImmutableData.deepFreeze(this._obj);
		}

	}

	get numbers() {
		return this._numbers;
	}

	get data() {
		return this._obj;
	}

	get date() {
		return new Date(this._date.valueOf());
	}

	/**
	 * Recursively freezes the given object's properties, when an object 
	 * is frozen an exception will be thrown if an property is modified.
	 * @param {any} obj The object to be deeply freezed.
	 */
	static deepFreeze(obj) {

		// console.log('deepFreeze: ', obj);

		if (obj === null) return;
		if (obj === undefined) return;

		for (let p of Object.keys(obj)) {

			// console.log('deepFreeze: prop: ', p, "object: ", obj[p])

			if (Array.isArray(obj[p])) {
				for(let i = 0; i < obj[p].length; i++)
					ImmutableData.deepFreeze(obj[p][i]);
				continue;
			}

			if (obj[p] === Object(obj[p])) {
				ImmutableData.deepFreeze(obj[p]);
			}

			Object.freeze(obj[p]);
		}

		Object.freeze(obj);

	}

}

File: app/util/ImmutableData.java
package util;

import java.util.List;
import java.util.Date;
import java.util.Arrays;
import java.util.Collections;

/**
 * 1) Implement an immutable class in Java. The class fully encapsulates an `int[]`, a `Collection` of `String`, 
 * and a `java.util.Date`. It should allow clients to specify all three when the class is constructed, 
 * and it includes accessors for each. The value returned by each accessor should not change after 
 * the object is instantiated.
 */
public final class ImmutableData {

	private final int numbers[];

	private final List<String> strings;

	private final Date date;

	/**
	 * 
	 */
	public ImmutableData(final int[] numbers, final List<String> strings, final Date date) {

		this.numbers = new int[numbers.length];
		System.arraycopy(numbers, 0, this.numbers, 0, numbers.length);

		this.strings = Arrays.asList(strings.toArray(new String[strings.size()]));
		this.date = date;

	}

	public Date getDate() {
		return (Date)date.clone();
	}

	public int[] getNumbers() {
		final int[] nums = new int[numbers.length];
		System.arraycopy(numbers, 0, nums, 0, numbers.length);
		return nums;
	}

	public List<String> getStrings() {
		return Arrays.asList(strings.toArray(new String[strings.size()]));
	}

}